cmd/restic/cmd_ls.go

```diff
@@ -59,19 +71,55 @@ func runLs(opts LsOptions, gopts GlobalOptions, args []string) error {
 		return err
 	}
 
+	// extract any specific directories to walk
+	var dirs []string
+	if len(args) > 1 {
+		dirs = args[1:]
+	}
+
 	ctx, cancel := context.WithCancel(gopts.ctx)
 	defer cancel()
-	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args) {
-		Verbosef("snapshot %s of %v at %s):\n", sn.ID().Str(), sn.Paths, sn.Time)
+	for sn := range FindFilteredSnapshots(ctx, repo, opts.Host, opts.Tags, opts.Paths, args[:1]) {
+		dirsToPrint := opts.Paths
+		if len(dirs) > 0 {
+			dirsToPrint = dirs
+		}
+		Verbosef("snapshot %s of %v at %s):\n", sn.ID().Str(), dirsToPrint, sn.Time)
 
 		err := walker.Walk(ctx, repo, *sn.Tree, nil, func(nodepath string, node *restic.Node, err error) (bool, error) {
 			if err != nil {
 				return false, err
 			}
-
 			if node == nil {
 				return false, nil
 			}
+
+			// apply any directory filters
+			if len(dirs) > 0 {
+				var nodeDir string
+				if !opts.Recursive {
+					// only needed for exact directory match; i.e. no subfolders
+					nodeDir = filepath.Dir(nodepath)
+				}
+				var match bool
+				for _, dir := range dirs {
+					if opts.Recursive {
+						if strings.HasPrefix(nodepath, dir) {
+							match = true
+							break
+						}
+					} else {
+						if nodeDir == dir {
+							match = true
+							break
+						}
+					}
+				}
+				if !match {
+					return true, nil
```

Comment:
Ah, this may be confusing in the docs: When the walker function returns `ignore=true`, the **node** won't be visited again, which is identified by the tree ID, which may be the same for different directories at different paths. For example, consider a backup of a couple of empty dirs, e.g. `/home/foo/dir1` and `/other/dir2`, and calling `restic ls latest /other`. When the walker runs the function for `/home/foo/dir1` and `ignore=true` is returned (because it is not to be listed), it will add the tree ID for the empty directory to its list of trees to ignore, and the walker function will not be called again.

Instead, return `ignore=false` and return `walker.SkipNode` as the `error` instead.

I'm sorry that it's so confusing, the interface was adapted from `filepath.Walk`, and later I added the option to not only ignore stuff based on the path (return `SkipNode`) but also based on the tree ID (return `ignore=true`). Sigh. If you have an idea on how to improve the docs, please let me know!

Reply:
Ohhh... interesting! Thanks for clarifying that for me.

I wonder, in hindsight, is my usage of these return values correct in these places?

- https://github.com/restic/restic/blob/abc923f69391ae55b0319c31ff2560083c874322/cmd/restic/cmd_stats.go#L177-L184 (I noticed that in cmd_ls.go, you return false instead of true in these cases.)
- https://github.com/restic/restic/blob/abc923f69391ae55b0319c31ff2560083c874322/cmd/restic/cmd_stats.go#L236 (now I'm not sure if I didn't want to visit the node again or if I wanted to not visit the tree again)

Update: However, when I change this line to `return false, walker.SkipNode`, the tree is not walked at all. The ls does in fact work correctly with the line as-is... 🤔

Reply:
I'm not sure what you mean by "this line", can you clarify?

Reply:
FYI, the walker has many tests for all cases, maybe that can clarify it a bit, see:

https://github.com/restic/restic/blob/abc923f69391ae55b0319c31ff2560083c874322/internal/walker/walker_test.go#L216-L224

Reply:
It doesn't, have a look at this repo: https://fd0.me/tmp/repo.tar.gz (the password is `foo`):

```
$ ./restic ls -l latest
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [] at 2018-08-11 22:09:31.33743769 +0200 CEST):
drwxr-xr-x  1000   100      0 2018-08-11 22:01:11 /test
drwxr-xr-x  1000   100      0 2018-08-11 22:02:58 /test/bar
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/bar/test.txt
drwxr-xr-x  1000   100      0 2018-08-11 22:04:19 /test/foo
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/foo/test.txt


$ ./restic ls -l latest /test/bar
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [/test/bar] at 2018-08-11 22:09:31.33743769 +0200 CEST):
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/bar/test.txt

$ ./restic ls -l latest /test/foo
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [/test/foo] at 2018-08-11 22:09:31.33743769 +0200 CEST):
```

The third call of `restic ls` did not print anything, although it should have printed the contents of the `test/foo` directory. The reason is that `/test/foo` and `/test/bar` have the same tree, which is on the ignore list because when `Walk()` processed `/test/bar`, it did not match the filter.

By the way, I think the directory itself (e.g. `/test/bar`) should also be printed.

Reply:
> The SkipNode value seems very aggressive in what it skips... I'm still working on it though.

The semantics are modeled after `filepath.Walk()`...

> What do you mean by this? Which directory exactly, and printed where? 

I meant that if I run `restic ls latest /test/bar` then I think it's good to have the directory node `/test/bar` itself also printed, like this:
```
repository 1271b4d6 opened successfully, password is correct
found 1 old cache directories in /home/fd0/.cache/restic, pass --cleanup-cache to remove them
snapshot 62bbf4fb of [/test/bar] at 2018-08-11 22:09:31.33743769 +0200 CEST):
drwxr-xr-x  1000   100      0 2018-08-11 22:02:58 /test/bar
-rw-r--r--  1000   100      0 2018-08-11 22:02:58 /test/bar/test.txt
```

What do you think?